lib/core/include/scipp/core/element/arithmetic.h

```diff
@@ -87,25 +88,72 @@ struct subtract_types_t {
-                     std::tuple<std::tuple<Eigen::Vector3d, int64_t>>(),
-                     std::tuple<std::tuple<Eigen::Vector3d, int32_t>>()));
+  using types = decltype(std::tuple_cat(
+      std::declval<arithmetic_type_pairs_with_bool>(),
+      std::tuple<std::tuple<double, Eigen::Vector3d>>(),
+      std::tuple<std::tuple<float, Eigen::Vector3d>>(),
+      std::tuple<std::tuple<int64_t, Eigen::Vector3d>>(),
+      std::tuple<std::tuple<int32_t, Eigen::Vector3d>>(),
+      std::tuple<std::tuple<Eigen::Vector3d, double>>(),
+      std::tuple<std::tuple<Eigen::Vector3d, float>>(),
+      std::tuple<std::tuple<Eigen::Vector3d, int64_t>>(),
+      std::tuple<std::tuple<Eigen::Vector3d, int32_t>>(),
+
+      std::tuple<std::tuple<Eigen::Affine3d, scipp::core::RotationTransform>>(),
+      std::tuple<std::tuple<Eigen::Affine3d, scipp::core::ScalingTransform>>(),
+      std::tuple<std::tuple<Eigen::Affine3d, Eigen::Matrix3d>>(),
+
+      std::tuple<std::tuple<scipp::core::TranslationTransform,
+                            scipp::core::RotationTransform>>(),
+      std::tuple<std::tuple<scipp::core::TranslationTransform,
+                            scipp::core::ScalingTransform>>(),
+      std::tuple<
+          std::tuple<scipp::core::TranslationTransform, Eigen::Matrix3d>>(),
+
+      std::tuple<std::tuple<scipp::core::RotationTransform, Eigen::Affine3d>>(),
+      std::tuple<std::tuple<scipp::core::RotationTransform, Eigen::Vector3d>>(),
+      std::tuple<std::tuple<scipp::core::RotationTransform,
+                            scipp::core::TranslationTransform>>(),
+      std::tuple<std::tuple<scipp::core::RotationTransform,
+                            scipp::core::RotationTransform>>(),
+      std::tuple<std::tuple<scipp::core::RotationTransform,
+                            scipp::core::ScalingTransform>>(),
+      std::tuple<std::tuple<scipp::core::RotationTransform, Eigen::Matrix3d>>(),
+
+      std::tuple<std::tuple<scipp::core::ScalingTransform, Eigen::Affine3d>>(),
+      std::tuple<std::tuple<scipp::core::ScalingTransform, Eigen::Vector3d>>(),
+      std::tuple<std::tuple<scipp::core::ScalingTransform,
+                            scipp::core::TranslationTransform>>(),
+      std::tuple<std::tuple<scipp::core::ScalingTransform,
+                            scipp::core::RotationTransform>>(),
+      std::tuple<std::tuple<scipp::core::ScalingTransform,
+                            scipp::core::ScalingTransform>>(),
+      std::tuple<std::tuple<scipp::core::ScalingTransform, Eigen::Matrix3d>>(),
+
+      std::tuple<std::tuple<Eigen::Matrix3d, Eigen::Affine3d>>(),
+      std::tuple<std::tuple<Eigen::Matrix3d, Eigen::Vector3d>>(),
+      std::tuple<
+          std::tuple<Eigen::Matrix3d, scipp::core::TranslationTransform>>(),
+      std::tuple<std::tuple<Eigen::Matrix3d, scipp::core::RotationTransform>>(),
+      std::tuple<std::tuple<Eigen::Matrix3d, scipp::core::ScalingTransform>>(),
```

Comment:
Except for the `Vector3d` cases, this looks like all possible permutations? Have a look at the helpers in `core/transform_common.h` to define these in a more concise manner. `pair_product_t` is probably doing what you need.